pandas/core/construction.py

```diff
@@ -596,6 +597,8 @@ def sanitize_array(
         # create an extension array from its dtype
         _sanitize_non_ordered(data)
         cls = dtype.construct_array_type()
+        if isinstance(data, KeysView):
+            data = list(data)
```

Comment:
I think this is the critical part that is indeed fixing the issue (for `Series(..)` at least), and so this is a good change. 

I would just make the `if` check more generic. Because while the example was using `dict.keys()`, you also have other iterables (e.g. the `dict.values()`) that will have the same problem, and we should try to fix it for all of them.

Looking at the logic just below for the non-ExtensionDtype cases, there is a `if hasattr(data, "__array__")` and then after that in the final `else` it is also doing a `data = list(data)`. So maybe the check above could be `if not hasattr(data, "__array__")` instead of `if isinstance(data, KeysView)`.

Reply:
Oh makes sense! Will do!